cpp/src/arrow/io/hdfs-io-test.cc

```diff
@@ -0,0 +1,277 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gtest/gtest.h"
+
+#include <cstdlib>
+#include <sstream>
+#include <string>
+
+#include "arrow/io/hdfs.h"
+#include "arrow/test-util.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+namespace io {
+
+std::vector<uint8_t> RandomData(int64_t size) {
+  std::vector<uint8_t> buffer(size);
+  test::random_bytes(size, 0, buffer.data());
+  return buffer;
+}
+
+class TestHDFSClient : public ::testing::Test {
+ public:
+  void MakeScratchDir() {
+    if (client_->Exists(scratch_dir_)) {
+      ASSERT_OK(client_->Delete(scratch_dir_, true));
+    }
+    ASSERT_OK(client_->CreateDirectory(scratch_dir_));
+  }
+
+  void WriteDummyFile(const std::string& path, const uint8_t* buffer,
+      int64_t size, bool append=false, int buffer_size=0, int replication=0,
+      int default_block_size=0) {
+    std::shared_ptr<HDFSWriteableFile> file;
+    EXPECT_OK(client_->OpenWriteable(path, append, buffer_size, replication,
```

Comment:
I ran into a weird compilation error and changed these to EXPECT temporarily, but changing them back to ASSERT works now.